File: src/CsharpClient/Quix.Sdk.Streaming/Models/StreamWriter/EventDataBuilder.cs
﻿using Quix.Sdk.Process.Models;
using System;
using System.Collections.Generic;
using System.Text;

namespace Quix.Sdk.Streaming.Models.StreamWriter
{
    /// <summary>
    /// Builder for creating <see cref="EventData"/> packages within the <see cref="StreamEventsWriter"/> 
    /// </summary>
    public class EventDataBuilder
    {
        private readonly StreamEventsWriter streamEventsWriter;
        private readonly long timestampNanoseconds;

        private readonly List<Process.Models.EventDataRaw> events = new List<Process.Models.EventDataRaw>();
        private readonly Dictionary<string, string> tags;

        /// <summary>
        /// Initializes a new instance of <see cref="EventDataBuilder"/>
        /// </summary>
        /// <param name="streamEventsWriter">Events writer owner</param>
        /// <param name="timestampNanoseconds">Timestamp assigned to the Events created by the builder</param>
        public EventDataBuilder(StreamEventsWriter streamEventsWriter, long timestampNanoseconds)
        {
            this.streamEventsWriter = streamEventsWriter;
            this.timestampNanoseconds = timestampNanoseconds;
            this.tags = new Dictionary<string, string>(this.streamEventsWriter.DefaultTags);
        }

        /// <summary>
        /// Adds new event value at the time the builder is created for
        /// </summary>
        /// <param name="eventId">Event Id</param>
        /// <param name="value">Value of the event</param>
        /// <returns></returns>
        public EventDataBuilder AddValue(string eventId, string value)
        {
            var @event = new Process.Models.EventDataRaw()
            {
                Id = eventId,
                Timestamp = timestampNanoseconds,
                Value = value
            };

            this.events.Add(@event);

            return this;
        }

        /// <summary>
        /// Adds a tag to the event values
        /// </summary>
        /// <param name="tagId">Tag Id</param>
        /// <param name="value">Tag value</param>
        public EventDataBuilder AddTag(string tagId, string value)
        {
            this.tags[tagId] = value;

            return this;
        }
        
        /// <summary>
        /// Adds tags to the event values.
        /// </summary>
        /// <param name="tagsValues">Tags values.</param>
        public EventDataBuilder AddTags(IEnumerable<KeyValuePair<string, string>> tagsValues)
        {
            if (tagsValues == null) return this;
            
            foreach (var tag in tagsValues)
            {
                this.tags[tag.Key] = tag.Value;
            }

            return this;
        }

        /// <summary>
        /// Writes the events
        /// </summary>
        public void Write()
        {
            // Tags
            foreach(var ev in this.events)
            {
                ev.Tags = this.tags;
            }

            this.streamEventsWriter.Write(this.events);
        }

    }

}

File: src/PythonClient/lib/quixstreaming/builders/eventdatabuilder.py
from typing import Dict

from quixstreaming import __importnet
import Quix.Sdk.Streaming
import clr
clr.AddReference('System.Collections')
import System.Collections.Generic

class EventDataBuilder(object):
    """
        Builder for creating event data packages for StreamPropertiesWriter
    """

    def __init__(self, net_object: Quix.Sdk.Streaming.Models.StreamWriter.EventDataBuilder):
        """
            Initializes a new instance of EventDataBuilder.

            Events:

            net_object (.net object): The .net object representing EventDataBuilder.
        """

        self.__wrapped = net_object

    def add_value(self, event_id: str, value: str) -> 'EventDataBuilder':
        """
        Adds new event at the time the builder is created for
        :param event_id: The id of the event to set the value for
        :param value: the string value
        :return: The builder
        """
        self.__wrapped = self.__wrapped.AddValue(event_id, value)
        return self

    def add_tag(self, tag_id: str, value: str) -> 'EventDataBuilder':
        """
        Sets tag value for the values
        :param tag_id: The id of the tag
        :param value: The value of the tag
        :return: The builder
        """
        self.__wrapped = self.__wrapped.AddTag(tag_id, value)
        return self

    def add_tags(self, tags: Dict[str, str]) -> 'EventDataBuilder':
        """
            Copies the tags from the specified dictionary. Conflicting tags will be overwritten
            :param tags: The tags to add

        :return: EventDataBuilder
        """

        if tags is None:
            return self

        prep_tags_dict = System.Collections.Generic.List[System.Collections.Generic.KeyValuePair[str, str]]([])
        for key, val in tags.items():
            prep_tags_dict.Add(System.Collections.Generic.KeyValuePair[str, str](key, val))
        self.__wrapped.AddTags(prep_tags_dict)
        return self

    def write(self):
        """
        Writes the values to the StreamEventsWriter buffer. See StreamEventsWriter buffer settings for more information when the values are sent to the broker
        """
        self.__wrapped.Write()
